core/payments/reservation/reservation_ledger_cache.go

```diff
@@ -0,0 +1,206 @@
+	// monitors the PaymentVault for changes, and updates cached ledgers accordingly
+	vaultMonitor *ReservationVaultMonitor
+}
+
+func NewReservationLedgerCache(
+	ctx context.Context,
+	logger logging.Logger,
+	maxLedgers int,
+	paymentVault payments.PaymentVault,
+	timeSource func() time.Time,
+	overfillBehavior OverfillBehavior,
+	bucketCapacityPeriod time.Duration,
+	updateInterval time.Duration,
+) (*ReservationLedgerCache, error) {
+	if paymentVault == nil {
+		return nil, errors.New("payment vault must be non-nil")
+	}
+
+	if timeSource == nil {
+		return nil, errors.New("time source must be non-nil")
+	}
+
+	if bucketCapacityPeriod <= 0 {
+		return nil, errors.New("bucket capacity period must be > 0")
+	}
+
+	cache, err := lru.NewWithEvict(
+		maxLedgers,
+		func(accountAddress gethcommon.Address, reservationLedger *ReservationLedger) {
+			if !reservationLedger.IsBucketEmpty(timeSource()) {
+				logger.Errorf("evicted account %s from LRU reservation ledger cache, but the underlying leaky bucket "+
+					"wasn't empty! You must increase the ReservationLedgerCache LRU cache size", accountAddress.Hex())
+				return
+			}
+
+			logger.Debugf("evicted account %s from LRU reservation ledger cache", accountAddress.Hex())
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new LRU cache with evict: %w", err)
+	}
+
+	minNumSymbols, err := paymentVault.GetMinNumSymbols(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("get min num symbols: %w", err)
+	}
+
+	ledgerCache := &ReservationLedgerCache{
+		logger:               logger,
+		cache:                cache,
+		paymentVault:         paymentVault,
+		timeSource:           timeSource,
+		overfillBehavior:     overfillBehavior,
+		bucketCapacityPeriod: bucketCapacityPeriod,
+		minNumSymbols:        minNumSymbols,
+		ledgerCreationLock:   common.NewIndexLock(256),
+	}
+
+	// Create the vault monitor with callback functions
+	ledgerCache.vaultMonitor, err = NewReservationVaultMonitor(
+		ctx,
+		logger,
+		paymentVault,
+		updateInterval,
+		// relatively arbitrary value. much higher than account number in practice, but much lower than what the RPC
+		// could actually handle. Since the "sweet spot" is really wide, hardcode this instead of spending time wiring
+		// in a config value
+		1024,
+		ledgerCache.GetAccountsToUpdate,
+		ledgerCache.UpdateReservation,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new reservation vault monitor: %w", err)
+	}
+
+	return ledgerCache, nil
+}
+
+// GetOrCreate retrieves an existing ReservationLedger for the given account, or creates a new one if it doesn't exist
+func (c *ReservationLedgerCache) GetOrCreate(
+	ctx context.Context,
+	accountID gethcommon.Address,
+) (*ReservationLedger, error) {
+	// Fast path: check if ledger already exists in cache
+	if ledger, exists := c.cache.Get(accountID); exists {
+		return ledger, nil
+	}
+
+	// Slow path: acquire per-account lock and check again
+	accountIndex := binary.BigEndian.Uint64(accountID.Bytes()[:8])
```

Comment:
nit: constructor for IndexLock takes a uint32 but then its lock/unlock methods take a uint64. Feels inconsistent? Also not important but we could consider adding a bunch more typed lock/unlock methods on the accountIndex, such that we could pass in a geth-address, or a bigInt, etc. thoughts?

Reply:
> Feels inconsistent?

The total number of indices will exceed the number of locks, probably by a lot (otherwise, why use the index lock at all?). With that in mind, having the index be a larger int type seems reasonable to me.

> we could consider adding a bunch more typed lock/unlock methods on the accountIndex, such that we could pass in a geth-address, or a bigInt, etc.

Hmm, interesting point. Maybe makes sense? On the one hand, it's kinda nice that the existing index lock impl is super simple and pure.... index in, lock out. On the other hand, it's unsightly to be manually converting geth addresses to uint64 indices all over the place.

cc @cody-littley for discussion on the index lock util